File: src/2018/day02/python/solution.py
from collections import Counter


def format_input(input: str) -> list[str]:
    return [line.strip() for line in input.strip().splitlines() if line]


def count_line(line: str) -> tuple[int, int]:  # (3Count, 2Count)
    counter = Counter(line)
    counts = (0, 0)
    if 3 in counter.values():
        counts = (1, counts[1])
    if 2 in counter.values():
        counts = (counts[0], 1)

    return counts


def part1(input: str) -> int:
    ids = format_input(input)
    counted_ids = [count_line(id) for id in ids]
    count3, count2 = zip(*counted_ids)

    return sum(count3) * sum(count2)


def get_duplicate(ids: list[str]) -> str:
    seen = set()

    for id in ids:
        if id in seen:
            return id

        seen.add(id)

    return ""


def part2(input):
    ids = format_input(input)

    for replace_index in range(len(ids[0])):
        subbed_ids = [id[0:replace_index] + id[replace_index + 1 :] for id in ids]
        duplicate = get_duplicate(subbed_ids)
        if duplicate:
            return duplicate

    return ""

File: src/2018/day02/typescript/solution.ts
function formatInput(input: string): string[] {
  return input.trim().split('\n');
}

function countLine(line: string): [number, number] {
  const counts = new Map<string, number>();
  for (const c of line) {
    counts.set(c, (counts.get(c) ?? 0) + 1);
  }
  const values = [...counts.values()];
  const hasThree = values.includes(3);
  const hasTwo = values.includes(2);

  return [hasThree ? 1 : 0, hasTwo ? 1 : 0];
}

function getDuplicate(ids: string[]): string {
  const seen = new Set<string>();

  for (const id of ids) {
    if (seen.has(id)) {
      return id;
    }
    seen.add(id);
  }
  return '';
}

export function part1(input: string): number {
  const ids = formatInput(input);
  const countedIds = ids.map((line) => countLine(line));
  const sums = countedIds.reduce(
    (acc, count) => {
      acc[0] += count[0];
      acc[1] += count[1];
      return acc;
    },
    [0, 0],
  );
  return sums.reduce((a, v) => a * v);
}

export function part2(input: string): string {
  const ids = formatInput(input);

  for (let i = 0; i < ids[0].length; i++) {
    const subbedIds = ids.map((id) => id.slice(0, i) + id.slice(i + 1));
    const duplicate = getDuplicate(subbedIds);
    if (duplicate !== '') {
      return duplicate;
    }
  }

  return '';
}
